internal/common/regions.go

```diff
@@ -1,23 +1,112 @@
 package common
 
 import (
+	"context"
+	"fmt"
+	"strings"
+
+	"log"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
 	"github.com/aws/aws-sdk-go/aws/endpoints"
 )
 
-func GetAWSRegions(selectedRegions []string) []string {
+func GetAWSRegions(ctx context.Context, cfg aws.Config, selectedRegions []string) ([]string, error) {
+	log.Println("Starting GetAWSRegions function")
+
+	var regionsToCheck []string
 	if len(selectedRegions) > 0 {
-		return selectedRegions
+		regionsToCheck = selectedRegions
+		log.Printf("Using selected regions: %v\n", regionsToCheck)
+	} else {
+		log.Println("No regions selected, checking all regions")
+		resolver := endpoints.DefaultResolver()
+		partitions := resolver.(endpoints.EnumPartitions).Partitions()
+
+		for _, p := range partitions {
+			for region := range p.Regions() {
+				regionsToCheck = append(regionsToCheck, region)
+			}
+		}
+		log.Printf("All regions to check: %v\n", regionsToCheck)
 	}
 
-	resolver := endpoints.DefaultResolver()
-	partitions := resolver.(endpoints.EnumPartitions).Partitions()
+	// Find an enabled region
+	log.Println("Attempting to find an enabled region")
+	var enabledRegion string
+	var expiredToken bool
+	invalidTokenErrors := []string{}
+	for _, region := range regionsToCheck {
+		log.Printf("Checking region: %s\n", region)
+		testCfg := cfg.Copy()
+		testCfg.Region = region
+		stsClient := sts.NewFromConfig(testCfg)
+		_, err := stsClient.GetCallerIdentity(ctx, &sts.GetCallerIdentityInput{})
+		if err == nil {
+			enabledRegion = region
+			log.Printf("Found enabled region: %s\n", enabledRegion)
+			break
+		} else {
+			errMsg := err.Error()
+			if strings.Contains(errMsg, "ExpiredToken") {
+				expiredToken = true
+				log.Printf("Token is expired: %v\n", err)
+				break
+			} else if strings.Contains(errMsg, "InvalidClientTokenId") {
+				invalidTokenErrors = append(invalidTokenErrors, fmt.Sprintf("Region %s: %s", region, errMsg))
+				log.Printf("Token is invalid for region %s: %v\n", region, err)
+			} else if strings.Contains(errMsg, "no such host") {
+				log.Printf("Region %s is not accessible: %v\n", region, err)
+			} else {
+				log.Printf("Region %s is not enabled: %v\n", region, err)
+			}
+		}
+	}
+
+	if expiredToken {
+		return nil, fmt.Errorf("the AWS token has expired")
```

Comment:
Why would you do this here and not above where you check if we have an expired token?